### DISCUSSION TITLE
How to inject CSLA classes into a class library of a Quartz project?

### DISCUSSION THREAD

**Discussion by ecaldentey** [MarimerLLC/csla#3923] [Questions] [answered]
Hello everyone,

I have a Quartz Server project where I register my Csla interfaces and another project, which is a class library where I define the jobs that execute. This latter project has many classes, and I'm not sure how to inject a property as I do in Blazor Razor pages, something like:

```
@inject IDataPortal<Library.ProjectEdit> FactoryProject

private async Task GetGrandfather()
{
var dataportal = new Library.ProjectEdit.Factory(FactoryProject);
ProjectDto projectDto = new ProjectDto
{
projectDto = 1
};

csharp
Copy code
var myProject = await dataportal.GetAsync(projectDto);
}
```

Any help on how to inject the Csla classes into a class library of a Quartz project is appreciated.

**Comment by rockfordlhotka** [comment]
This is probably a question you'll need to ask the Quartz Server users or support: how do I inject a dependency into a job or task class?

I had never heard of Quartz Server until today when I looked it up for this question.

I'm going to guess that when you create a job, that job is a class? Hopefully they support DI, and allow you to do something like this:

```c#
  public class MyJob
  {
    public MyJob(IDataPortal<ProjectEdit> portal)
    {
      _portal = portal;
    }

    private IDataPortal<ProjectEdit> _portal;

    public async Task DoTheJob()
    {
      var project = await _portal.FetchAsync(123);
    {
  }
```

**Reply by ecaldentey** [reply]
Hello  Rocky

This approach doesn't work because IDataPortal<ProjectEdit> portal is Scoped and the job's constructor only supports services declared as Singleton.

Doesn't work:
public MyJob(IDataPortal<ProjectEdit> portal)
{
    _portal = portal;
}

Works:
public JobDownloader(IConfiguration configuration, ILanguageService languageService)
{
    _configuration = configuration;
    _languageService = languageService;
}
While I'm resolving my issue, could you please indicate how I can manually create a dataportal in a class of a class library in a Windows project like Quartz? I need to manually create an instance FactoryProject to be able to execute 
var myProject = await dataportal.GetAsync(projectDto);

var dataportal = new Library.ProjectEdit.Factory(FactoryProject);
ProjectDto projectDto = new ProjectDto
{
    projectDto = 1
};
var myProject = await dataportal.GetAsync(projectDto);

**Reply by StefanOssendorf** [reply]
I think the way to go for you is to receive the `IServiceProvider` and create a scope on demand from which you can resolve the Dataportal.

**Reply by rockfordlhotka** [reply]
@StefanOssendorf beat me to the answer I was going to suggest 😎

**Reply by ecaldentey** [reply]
Hello @StefanOssendorf ,

Your response works for me, but the way I'm doing it, I'm directly accessing the methods of the interface, skipping all the CSLA implementation that constructs the BusinessClass and applies all the rules.

![image](https://github.com/MarimerLLC/csla/assets/65839718/2fd93a76-f8f2-402a-abb0-4616fb2450eb)

public JobDownloader(IConfiguration configuration, ILanguageService languageService, IServiceProvider serviceProvider)
{
    _configuration = configuration;
    _languageService = languageService;
    _serviceProvider = serviceProvider;
}

public async Task Execute(IJobExecutionContext context)
{
    using (var scope = _serviceProvider.CreateScope())
        {
            var dataportal = scope.ServiceProvider.GetService<IProjectDal>();

            ProjectDto projectDto = new ProjectDto
            {
                ProjectId = 1
            };

            var project = dataportal.FetchObject(projectDto, ProjectSelectDbTypes.SelectByKey, "");
            
        }
}

Can you tell me what I'm doing wrong? Because what I need is to have access to the methods of the factory to create CSLA classes.

Thanks in advance

**Reply by rockfordlhotka** [reply]
Are you able to call UseCsla during app startup? That is necessary to set up all the services required by CSLA, including the data portal.

Otherwise you will have to look at the CSLA source sources and walk through the entire call chain to create the client-side and server-side services necessary for the complete data portal pipeline on logical client and server.

That should be totally doable, just tedious.

**Reply by ecaldentey** [reply, marked as answer]
@rockfordlhotka,

After making all the service registrations, I added this line of code:
builder.Services.AddCsla();

And magically, the following line worked and stopped returning null:

var cslaFactory = scope.ServiceProvider.GetService<IDataPortal<ProjectEdit>>();
Thank you very much for the help.

**Comment by kcabral817** [comment]
I think you'll need either Spring or Google Guice for JAVA (or some other DI framework) then use the @Inject attribute on the property you want to inject.  But I think you'll need to do this in JAVA for Quartz since it's a JAVA based job scheduler.  But the information I'm giving you should be considered a WAG.

**Reply by ecaldentey** [reply]
I'm going to try with Ninject to see if it works
